web/packages/teleterm/src/ui/services/workspacesService/workspacesService.ts

```diff
@@ -647,8 +617,92 @@ function getDocumentsToPersist(documents: Document[]): Document[] {
 }
 
 function getLocationToRestore(
-  documents: Document[],
+  documents: Immutable<Document[]>,
   location: DocumentUri
 ): DocumentUri | undefined {
   return documents.find(d => d.uri === location) ? location : documents[0]?.uri;
 }
+
+function getWorkspaceDefaultState(
+  rootClusterUri: RootClusterUri,
+  restoredWorkspace?: Immutable<Omit<Workspace, 'accessRequests'>>
+): Workspace {
+  const defaultDocument = createClusterDocument({ clusterUri: rootClusterUri });
+  const defaultWorkspace: Workspace = {
+    accessRequests: {
+      pending: getEmptyPendingAccessRequest(),
+      isBarCollapsed: false,
+    },
+    location: defaultDocument.uri,
+    documents: [defaultDocument],
+    connectMyComputer: undefined,
+    documentsRestoredOrDiscarded: false,
+    localClusterUri: rootClusterUri,
+    unifiedResourcePreferences: parseUnifiedResourcePreferences(undefined),
+  };
+  if (!restoredWorkspace) {
+    return defaultWorkspace;
+  }
+
+  defaultWorkspace.localClusterUri = restoredWorkspace.localClusterUri;
+  defaultWorkspace.unifiedResourcePreferences = parseUnifiedResourcePreferences(
+    restoredWorkspace.unifiedResourcePreferences
+  );
+  defaultWorkspace.connectMyComputer = restoredWorkspace.connectMyComputer;
+
+  return defaultWorkspace;
+}
+
+// TODO(gzdunek): Parse the entire workspace state read from disk like below.
+function parseUnifiedResourcePreferences(
+  unifiedResourcePreferences: unknown
+): UnifiedResourcePreferences | undefined {
+  try {
+    return unifiedResourcePreferencesSchema.parse(
+      unifiedResourcePreferences
+    ) as UnifiedResourcePreferencesSchemaAsRequired;
+  } catch (e) {
+    new Logger('WorkspacesService').error(
+      'Failed to parse unified resource preferences',
+      e
+    );
+  }
+}
+
+function hasDocumentsToReopen({
+  previousDocuments,
+  currentDocuments,
+  defaultDocuments,
+}: {
+  previousDocuments?: Immutable<Document[]>;
+  currentDocuments: Document[];
+  defaultDocuments: Document[];
+}): boolean {
+  const omitUriAndTitle = (documents: Immutable<Document[]>) =>
+    documents.map(d => ({ ...d, uri: undefined, title: undefined }));
+
+  if (!previousDocuments?.length) {
+    return false;
+  }
+
+  // If current documents are no longer the default documents, don't offer reopening the previous session.
+  // This can happen when another dialog closed the dialog to reopen documents, and then you opened
+  // a few documents in that workspace.
+  // The app would then ask you about replacing your current documents with the default document
+  // the next time you would navigate to the workspace.
+  // TODO(gzdunek): Consider replacing the document reopen dialog with an action in the UI that
+  // could reopen the previous session at any time.
+  if (
```

Comment:
I'm not too happy with this check but it was the only simple workaround that came to my mind.

Reply:
I feel like allowing this situation to happen (where documents-reopen gets closed by cluster-login to another workspace which is then closed) makes reasoning about `previousDocuments`/`currentDocuments` quite difficult. It adds another state those two can be in and the situation in which this can happen is in itself quite unusual.

We'd remove a lot of complexity if we didn't let that happen, but I don't think it makes sense to stack even more changes in this PR.

If it's not possible to set `rootClusterUri` to undefined for reasons described in https://github.com/gravitational/teleport/pull/50063#discussion_r1888726089, then perhaps another solution would be to make the deep link handler re-execute `setActiveWorkspace` with the workspace that was active at the time when the deep link handler was called.

Just like with many other changes regarding the startup flow, I'm not 100% sure if it won't break anything else, but it seems to make sense. 🙃

Reply:
>I feel like allowing this situation to happen (where documents-reopen gets closed by cluster-login to another workspace which is then closed) makes reasoning about previousDocuments/currentDocuments quite difficult. It adds another state those two can be in and the situation in which this can happen is in itself quite unusual.

Agree, the startup flow became even more complex with changes from this PR :/

>If it's not possible to set rootClusterUri to undefined for reasons described in https://github.com/gravitational/teleport/pull/50063#discussion_r1888726089, then perhaps another solution would be to make the deep link handler re-execute setActiveWorkspace with the workspace that was active at the time when the deep link handler was called.

I feel that rendering workspaces when rootClusterUri is undefined may require many changes and testing, and I'm afraid that I will spend too much time on it.

Re-executing setActiveWorkspace when we didn't manage to set the correct workspace feels really simple, and should solve the problem the problem with this additional check, since we will always end up setting a workspace. 
I can push a one more commit here that will fix it.

Reply:
Yeah definitely, I agree.

Reply:
Ok, I realized that the logic around `documentsRestoredOrDiscarded` was broken from the beginning, so that I need that additional check.
We can't verify if we have documents to reopen every time we change the workspace - it must happen at the moment we restore previous workspaces. Otherwise, you can start a session with the default document (so `documentsRestoredOrDiscarded: false`) and then open a new one. Then you change a workspace, and come back. In `setActiveWorkspace` we would check if you have something to restore and this condition:
```ts
!arrayObjectIsEqual(
    omitUriAndTitle(previousDocuments),
    omitUriAndTitle(currentDocuments)
  )
```
would say yes.

I just fixed this, I promise it was the last change in this PR!